httpcore/_backends/sync.py

```diff
@@ -17,6 +19,117 @@
 from .base import SOCKET_OPTION, NetworkBackend, NetworkStream
 
 
+class SyncTLSStream(NetworkStream):
+    """
+    Because the standard `SSLContext.wrap_socket` method does
+    not work for `SSLSocket` objects, we need this class
+    to implement TLS stream using an underlying `SSLObject`
+    instance in order to support TLS on top of TLS.
+    """
+
+    # Defined in RFC 8449
+    TLS_RECORD_SIZE = 16384
+
+    def __init__(
+        self,
+        sock: socket.socket,
+        ssl_context: ssl.SSLContext,
+        server_hostname: typing.Optional[str] = None,
+        timeout: typing.Optional[float] = None,
+    ):
+        self._sock = sock
+        self._incoming = ssl.MemoryBIO()
+        self._outgoing = ssl.MemoryBIO()
+
+        self.ssl_obj = ssl_context.wrap_bio(
+            incoming=self._incoming,
+            outgoing=self._outgoing,
+            server_hostname=server_hostname,
+        )
+
+        self._perform_io(self.ssl_obj.do_handshake, timeout)
+
+    def _perform_io(
+        self,
+        func: typing.Callable[..., typing.Any],
+        timeout: typing.Optional[float],
+    ) -> typing.Any:
+        ret = None
+        timeout = timeout or None  # Replaces `0` with `None`
+
+        while True:
+            errno = None
+            try:
+                ret = func()
+            except (ssl.SSLWantReadError, ssl.SSLWantWriteError) as e:
+                errno = e.errno
+
+            if timeout is not None and timeout <= 0:  # pragma: no cover
+                raise socket.timeout()
+
+            self._sock.settimeout(timeout)
+            operation_start = perf_counter()
+            self._sock.sendall(self._outgoing.read())
+            # If the timeout is `None`, don't touch it.
+            timeout = timeout and timeout - (perf_counter() - operation_start)
+
+            if errno == ssl.SSL_ERROR_WANT_READ:
+                if timeout is not None and timeout <= 0:  # pragma: no cover
+                    raise socket.timeout()
+
+                self._sock.settimeout(timeout)
+                operation_start = perf_counter()
+                buf = self._sock.recv(self.TLS_RECORD_SIZE)
+
+                # If the timeout is `None`, don't touch it.
```

Comment:
```suggestion
                # If the timeout is `None` or `Zero`, don't touch it.
```

Reply:
It can not be Zero

Reply:
It can be! check [this](https://github.com/encode/httpcore/pull/732/commits/8e38d22e3f06e5324076ef49f80c9fd86c5d4e11#diff-d3a40cf09f980d710bfc8d32385d11086ebd179cec344a9ac5c095251f7b6224R48) and [here](https://github.com/encode/httpcore/pull/732/files#diff-d3a40cf09f980d710bfc8d32385d11086ebd179cec344a9ac5c095251f7b6224R67)

Reply:
If the timeout is zero, then this part should raise an exception, am i wrong?

https://github.com/encode/httpcore/blob/92041d488ed10a48462711885255aa36b0bb429f/httpcore/_backends/sync.py#L67